resources/js/tableofcontent.js

```diff
@@ -1,3 +1,63 @@
+function isHeaderElement(elem) {
+    return elem.nodeName.match(/^H\d+$/) && elem.textContent;
+}
+
+function createHeaderData(elem, i) {
+    const headerText = elem.textContent.trim().replace(/"/g, '\\"');
+    const level = parseInt(elem.nodeName.substr(1));
+    const anchor = `kiwix-toc-${i}`;
+
+    const anchorElem = document.createElement("a");
+    anchorElem.id = anchor;
+
+    /* Mark header content with something we can reference. */
+    elem.insertAdjacentElement("afterbegin", anchorElem);
+
+    return { headerEntry: { text: headerText, anchor, table: [] }, level };
+}
+
+function getDOMHeadersPreorder(elem, array)
+{
+    if (isHeaderElement(elem))
+        array.push(elem);
+
+    Array.from(elem.children).forEach(child => getDOMHeadersPreorder(child, array));
+}
+
+function createTOCFromHeaders(headers, toc)
+{
+    /* Stack to keep track of parent header levels. */
+    const levelStack = [];
+    headers.forEach(function(elem, i) {
+        const { headerEntry, level } = createHeaderData(elem, i);
+
+        /* Pop until the correct level for insertion. */
+        while (levelStack.length !== 0 && level <= levelStack[levelStack.length - 1].level)
+            levelStack.pop();
+
+        if (levelStack.length === 0)
+            toc.table.push(headerEntry);
+        else
+            levelStack[levelStack.length - 1].table.push(headerEntry);
+
+        levelStack.push({ table: headerEntry.table, level });
+    }); 
+}
```

Comment:
What do you think about representing the TOC as a flat array of `{level, headerText, anchor}` objects and moving the conversion to hierarchical form to C++?

Reply:
It would likely be faster and JS array references are probably very confusing for C++ programmers...

I will do this in webview and cache it. I will also try to reduce the parsing cost by trying to see if webchannel can pass a JS object directly as JSON.

Reply:
Don't fall into the premature optimization trap. This is not the place where we should mind about performance too much. Simplicity of design and clarity of the code are more important. It is always beneficial having the simplest implementation in the history and then optimizing upon it if needed.

Reply:
Done